cpp/include/raft/core/operators.hpp

```diff
@@ -361,5 +372,24 @@ struct map_args_op {
   }
 };
 
+template <typename ComposedOpT>
+struct write_only_op {
+  const ComposedOpT composed_op;
+
+  template <typename OpT     = ComposedOpT,
+            typename UnusedT = std::enable_if_t<std::is_default_constructible_v<OpT>>>
+  constexpr write_only_op()
+    : composed_op{}  // The compiler complains if composed_op is not initialized explicitly
+  {
+  }
+  constexpr write_only_op(ComposedOpT o) : composed_op{o} {}
+
+  template <typename OutT, typename... Args>
+  constexpr RAFT_INLINE_FUNCTION void operator()(OutT* out, Args&&... args) const
+  {
+    *out = composed_op(std::forward<Args>(args)...);
+  }
+};
+
```

Comment:
I see, this operator should be useful for `linalg::writeOnlyUnaryOp`. Yet, I'd argue against adding this, and using the `writeOnlyUnaryOp` in general. The name of that helper is a little misleading and passing/dereferencing a pointer is unnecessary in many use-cases.  I've been using that helper for a while, because I was too lazy to propose an alternative that wouldn't need dereferencing a pointer.

Do you know if there exists a convenient helper in thrust to fill a buffer using an unary function on an index range? If not, maybe it's time to add something like this to raft (or just leave the current use-case a lambda)?

Reply:
I agree we could have a write-only helper a bit more along the lines of how unary op is working, i.e the op returns the value instead of taking a pointer and writing the output in memory. As long as `writeOnlyUnaryOp` is in use, the helper `write_only_op` can be useful IMO.

And yes, I believe we can also simply use `thrust::generate`:

```c++
template <typename DerivedPolicy,
          typename ForwardIterator,
          typename Generator>
__host__ __device__ void generate(
    const thrust::detail::execution_policy_base< DerivedPolicy > & exec,
    ForwardIterator first,
    ForwardIterator last,
    Generator gen);
```

Reply:
Actually [`thrust::tabulate`](https://nvidia.github.io/thrust/api/groups/group__transformations.html#function-tabulate) seems to do the job. `thrust::generate` is only for generators that don't take arguments (e.g constant, random)

Reply:
Btw the documentation for `write_only_unary_op` is incorrect:

https://github.com/rapidsai/raft/blob/9944b3a8b83bfd6cd8298a73cd175298e168d264/cpp/include/raft/linalg/unary_op.cuh#L126-L127

We have the following options:

- Change the behavior of `write_only_unary_op` and/or `writeOnlyUnaryOp` to actually take a functor with the signature `InType func(const InType& val);` (breaking change)
- Create another helper with a different name wrapping around `thrust::tabulate` (or our own custom kernel)
- Stop using `writeOnlyUnaryOp` and use `thrust::tabulate` directly.

cc @cjnolet, what do you think about this?